core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AdaptivePartitioning.java

```diff
@@ -0,0 +1,195 @@
+                        fragment.getId(), succinctBytes(estimatedMemoryConsumptionInBytes), partitionCount, runtimeAdaptivePartitioningPartitionCount);
+                return rewriteWith(new Rewriter(runtimeAdaptivePartitioningPartitionCount, context.idAllocator(), stageRuntimeInfoProvider), plan);
+            }
+        }
+
+        return plan;
+    }
+
+    public static boolean consumesHashPartitionedInput(PlanFragment fragment)
+    {
+        return isPartitioned(fragment.getPartitioning());
+    }
+
+    private static boolean isPartitioned(PartitioningHandle partitioningHandle)
+    {
+        return partitioningHandle.equals(FIXED_HASH_DISTRIBUTION) || partitioningHandle.equals(SCALED_WRITER_HASH_DISTRIBUTION);
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final int partitionCount;
+        private final PlanNodeIdAllocator idAllocator;
+        private final StageRuntimeInfoProvider runtimeInfoProvider;
+
+        private Rewriter(int partitionCount, PlanNodeIdAllocator idAllocator, StageRuntimeInfoProvider runtimeInfoProvider)
+        {
+            this.partitionCount = partitionCount;
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.runtimeInfoProvider = requireNonNull(runtimeInfoProvider, "runtimeInfoProvider is null");
+        }
+
+        @Override
+        public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
+        {
+            List<PlanNode> sources = node.getSources().stream()
+                    .map(context::rewrite)
+                    .collect(toImmutableList());
+            PartitioningScheme partitioningScheme = node.getPartitioningScheme();
+
+            PlanNodeId planNodeId = node.getId();
+            if (isEligibleRemoteExchange(node)) {
+                partitioningScheme = partitioningScheme.withPartitionCount(Optional.of(partitionCount));
+                planNodeId = idAllocator.getNextId();
+            }
+
+            return new ExchangeNode(
+                    planNodeId,
+                    node.getType(),
+                    node.getScope(),
+                    partitioningScheme,
+                    sources,
+                    node.getInputs(),
+                    node.getOrderingScheme());
+        }
+
+        private boolean isEligibleRemoteExchange(ExchangeNode exchangeNode)
+        {
+            if (exchangeNode.getScope() != REMOTE || exchangeNode.getType() != REPARTITION) {
+                return false;
+            }
+            PartitioningHandle partitioningHandle = exchangeNode.getPartitioningScheme().getPartitioning().getHandle();
+            return !partitioningHandle.isScaleWriters()
+                    && !partitioningHandle.isSingleNode()
+                    && partitioningHandle.getConnectorHandle() instanceof SystemPartitioningHandle;
```

Comment:
Does [this logic](https://github.com/trinodb/trino/pull/20380/files#diff-4de1971088a57bff022591dc19d8d759fda26b2d37a801d955e98b6f16d95118R340) apply here?